### PR TITLE
[DTensor][Bug Fix]Fix 2D DTensor mm with mesh_shape (1, n) or (n, 1)

### LINKED ISSUE #138742: Distributed Tensor raises error with torch 2.5
I'm encountering an IndexError when trying to perform a matrix multiplication between distributed tensors after upgrading to PyTorch 2.5.0. This same code was working perfectly in PyTorch 2.4.0.

Repro:
```python
import torch
from torch.distributed._tensor import init_device_mesh, Shard, distribute_tensor, Replicate

if __name__ == "__main__":
    mesh = init_device_mesh("cpu", (4,1))
    left = torch.randn(256, 128)
    right = torch.randn(128,256)
    
    left_d = distribute_tensor(left, mesh, [Shard(dim=0), Replicate()])
    right_d = distribute_tensor(right, mesh, [Replicate(), Replicate()])

    res = left_d @ right_d
    print(res)
    
```

On terminal:
```bash
torchrun --standalone --nnodes=1 --nproc-per-node=4 file.py
```

Throws:

[rank0]: Traceback (most recent call last):
[rank0]:   File "/Users/synnada/projects/composite-ml-new/file.py", line 12, in <module>
[rank0]:     res = left_d @ right_d
[rank0]:           ~~~~~~~^~~~~~~~~
[rank0]:   File "/Users/synnada/anaconda3/envs/py12/lib/python3.12/site-packages/torch/_compile.py", line 32, in inner
[rank0]:     return disable_fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/Users/synnada/anaconda3/envs/py12/lib/python3.12/site-packages/torch/_dynamo/eval_frame.py", line 632, in _fn
[rank0]:     return fn(*args, **kwargs)
[rank0]:            ^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/Users/synnada/anaconda3/envs/py12/lib/python3.12/site-packages/torch/distributed/tensor/_api.py", line 340, in __torch_dispatch__
[rank0]:     return DTensor._op_dispatcher.dispatch(
[rank0]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/Users/synnada/anaconda3/envs/py12/lib/python3.12/site-packages/torch/distributed/tensor/_dispatch.py", line 181, in dispatch
[rank0]:     self.redistribute_local_args(
[rank0]:   File "/Users/synnada/anaconda3/envs/py12/lib/python3.12/site-packages/torch/distributed/tensor/_dispatch.py", line 317, in redistribute_local_args
[rank0]:     resharded_local_tensor = redistribute_local_tensor(
[rank0]:                              ^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/Users/synnada/anaconda3/envs/py12/lib/python3.12/site-packages/torch/distributed/tensor/_redistribute.py", line 177, in redistribute_local_tensor
[rank0]:     transform_infos = _gen_transform_infos(current_spec, target_spec)
[rank0]:                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[rank0]:   File "/Users/synnada/anaconda3/envs/py12/lib/python3.12/site-packages/torch/distributed/tensor/_redistribute.py", line 100, in _gen_transform_infos
[rank0]:     target = target_placements[mesh_dim]
[rank0]:              ~~~~~~~~~~~~~~~~~^^^^^^^^^^
[rank0]: IndexError: list index out of range

### Versions

PyTorch version: 2.5.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.0 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: version 3.28.1
Libc version: N/A

Python version: 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 12:22:05) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-15.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] mypy==1.12.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==2.1.2
[pip3] torch==2.5.0
[pip3] torchaudio==2.3.1
[pip3] torchvision==0.18.1
[conda] numpy                     2.1.2                    pypi_0    pypi
[conda] torch                     2.5.0                    pypi_0    pypi
[conda] torchaudio                2.3.1                    pypi_0    pypi
[conda] torchvision               0.18.1                   pypi_0    pypi

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @tianyu-l @XilunWu

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139134

Fixes #138742. In the issue, the matrix multiplication with DTensor failed when the size of one of mesh dimension is 1 when the mesh is > 1D. We are missing tests for covering this corner case where mesh_shape is (n, 1) or (1, n). The DTensor mm op is correct when the 1D mesh is of shape (self.world_size, ) or 2D mesh with none of the mesh_dimension has a size of 1.  

In this PR, we fixed the corner case by updating `gen_einsum_strategies` in `_einsum_strategy.py`. Specifically, we cannot skip generating `mesh_dim_strategies` when `mesh_dim <= 1`, as this is not valid for nD mesh with one of the mesh dimension sizes being 1. 

Without the fix, the OpStrategy generated for 2D mesh with mesh_shape of (1,n) or (n,1) is wrong, as the OpStrategy generated is 1D. 

```
all_mesh_dim_strategies=[[[Replicate(), Replicate(), Replicate()], [Partial(sum), Shard(dim=1), Shard(dim=0)], [Shard(dim=0), Shard(dim=0), Replicate()], [Shard(dim=1), Replicate(), Shard(dim=1)]]]
OpStrategy(all_strategies):::   [(R, R) -> R, (S(1), S(0)) -> P, (S(0), R) -> S(0), (R, S(1)) -> S(1)] @ mesh: (4, 1)[(R, R) -> R, (S(1), S(0)) -> P, (S(0), R) -> S(0), (R, S(1)) -> S(1)] @ mesh: (4, 1)
```

After the fix, we can see the OpStrategy generated is correct with 2D strategy.
```
all_mesh_dim_strategies=[[[Replicate(), Replicate(), Replicate()], [Partial(sum), Shard(dim=1), Shard(dim=0)], [Shard(dim=0), Shard(dim=0), Replicate()], [Shard(dim=1), Replicate(), Shard(dim=1)]]][[[Replicate(), Replicate(), Replicate()], [Partial(sum), Shard(dim=1), Shard(dim=0)], [Shard(dim=0), Shard(dim=0), Replicate()], [Shard(dim=1), Replicate(), Shard(dim=1)]]] 
OpStrategy(all_strategies) = [(RR, RR) -> RR, (RS(1), RS(0)) -> RP, (RS(0), RR) -> RS(0), (RR, RS(1)) -> RS(1), (S(1)R, S(0)R) -> PR, (S(1)S(1), S(0)S(0)) -> PP, (S(1)S(0), S(0)R) -> PS(0), (S(1)R, S(0)S(1)) -> PS(1), (S(0)R, RR) -> S(0)R, (S(0)S(1), RS(0)) -> S(0)P, (S(0)S(0), RR) -> S(0)S(0), (S(0)R, RS(1)) -> S(0)S(1), (RR, S(1)R) -> S(1)R, (RS(1), S(1)S(0)) -> S(1)P, (RS(0), S(1)R) -> S(1)S(0), (RR, S(1)S(1)) -> S(1)S(1)] @ mesh: (4, 1)
```

*******
As a follow up, we should add more test coverage for DTensor op with 2D mesh and 2D mesh with one of the size of mesh dimension being 1. 
*******

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wconstab @d4l3k @c-p-i-o @tianyu-l @XilunWu

### CODE REVIEW

**Comment by fegin** [comment]
The failing tests looks legit. We should fix them first.

**Comment by wz337** [comment]
> The failing tests looks legit. We should fix them first.
Seems there are some issues when using "gloo" backend. So just skip the non gpu tests now and will follow up by looking into DTensorTestbase.

File: test/distributed/_tensor/test_matrix_ops.py [hunk 1]
```diff
@@ -339,6 +340,31 @@ def test_scaled_dot_product_attention(self):
...
+        Test mm with DTensor with 2D mesh.
+        We need to add the test here since we only test 1D mesh in test_dtensor_ops.py.
+        Also, we added tests for the corner case where one of the 2D dimension is 1.
+
+        # TODO: we need to test more DTensor ops with 2D mesh, especially when 1 of the
+        mesh dimension of the 2D mesh is 1.
+        """
+        mesh_0 = init_device_mesh(self.device_type, (self.world_size // 2, 2))
+        mesh_1 = init_device_mesh(self.device_type, (self.world_size, 1))
+        mesh_2 = init_device_mesh(self.device_type, (1, self.world_size))
```
Comment (kwen2501) [inline test/distributed/_tensor/test_matrix_ops.py:356]: nit: better to put `init_device_mesh` in the loop?

**Review by fegin — APPROVED** [review]

**Pull request closed without merging**